### ISSUE TITLE
Add disk pressure and disk eviction

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#3524]
We have a concept of memory pressure, and memory eviction. In this issue I propose to have a similar concept of **disk space pressure** and disk eviction: When we are getting close to running out of space on a disk, we ask the various disk evictors to start saving space. For example, in a disk shared by data, commitlog and hints, the data, commitlog and hint evictors can:

1.  Data: Tell the compaction controller to increase the compaction cpu and disk share, to allow compaction to save space. May also apply other emergency tweaks to the compaction strategy - e.g., starting compaction to see which sstables are most redundant (see issue #2132), compacting together even few sstables than usual, compacting even single sstables, etc. Watch out not to do this in the classic STCS because this strategy may need huge temporary space during compaction, and just make the disk pressure worse. 

2. Commitlog: Increase the memory flush shares, so more memory can be flushed to disk and commitlog data removed. Note that the benefit of this is not obvious, because commitlog files are deleted but sstables are written. So it depends what is the relative size of commitlog vs. sstable. However, another benefit of replacing commitlog with sstable is that it might allow a compaction (with older data) to save some more space.

3. Hints: we can drop the oldest hint files. This may sound risky (we will need repair when a node is down for a very long time) but it may be the opposite: 1. This feature will allow us to keep hints for a very long time when disk space is *plenty* and not artificially limit it to say 10% of the disk. 2. This feature in the case of a real disk-space emergency will cause hints to be lost quickly (we can warn the db admin about it) but that's better than losing data when we run out of disk!

The idea here is to replace ad-hoc heuristics like "hints shouldn't take more than 10% of the disk" that we now have (see issue #3516 ) by something more robust.

**Comment by vladzcloudius** [comment]
@nyh I like the idea in general but I'd like to clarify a few points.

In the context of HH this idea (of a back-pressure/eviction) has been implemented from day 1: if we run out of space we stop the producer and leave the consumer (sender) open. Once hints are sent out the corresponding segments are deleted thereby freeing the disk space.

The only difference between this and what you are suggesting is that you want to delete the oldest segments unconditionally while in the current solution we virtually delete the newest segment (we simply don't let them to be created by blocking the producer).

In the context of HH it's not obvious to me which strategy is preferable since segments are not per-KS or per-CF - they are per destination node and may contain mutations to different tables. Therefore it's not necessarily that the newest segments are going to contain the latest mutations to all KS-es. While from the codding perspective the current approach is a lot easier to implement.

As for the indication of hints being dropped - there is a corresponding metrics that will be bumped up for every hint that we refused to store (we don't differentiate between different classes reasons for a hint drop at the moment but we can easily add this if needed).

One more thing that we will definitely have to implement if/when we have this is the "prioritization" of different disk space consumers. For instance, you really don't want to let HH consume as much space as it wants simply because it may eat up all your disk space very fast:  hints are stored in a commitlog-like structured files that are not compacted, and if some actively written replica goes down then you'll get a storm of hints to that node. This is exactly where this 10% limitation came from (the scenario above has been pointed out by Avi back then).

This means that you'll have to come up with something similar to that "10% voodoo", and it may end up to be very similar to what we have right now.

**Comment by bhalevy** [comment]
Cc @raphaelsc

**Comment by nyh** [comment]
> Other things we can control when reaching capacity are:
> 
>     * disabling auto-snapshot
>     * deleting existing auto-snapshots

These ideas a risky - if a user counts on auto-snapshot (I think it's a bad idea and badly implemented, but let's assume someone counts on it), if you disable it without them knowing, it can lead to a disaster. 

But a discussion with @yarongilor reminding me of something else that can and should be stopped when reaching maximum disk capacity: **view building**. The background process of "view building" - which happens after you create a view on existing data - can easily be paused when running out of space. This won't cause any API breakages, because view building is always known to take a long amount of time so it's not a problem if it takes an even longer time.

The disk pressure controller would pause the view building when too little disk space is available, and resume it when disk space is plenty again (e.g., after the cluster grows).

**Comment by mykaul** [comment]
@nyh - the view building is a good idea (probably should be a separate item). Is it pause or stop?

**Comment by denesb** [comment]
Based on feedback from the plan review, we won't implement any of the more sophisticated algorithms suggested here, instead we will simply reject everything that generates writes when disk usage reaches certain percentage. With the sole exception of those writes that are needed to advance the group0 state machine, so that tablet migrations can proceed.

**Comment by nyh** [comment]
> Based on feedback from the plan review, we won't implement any of the more sophisticated algorithms suggested here, instead we will simply reject everything that generates writes when disk usage reaches certain percentage.

If I understand you correctly, your suggestion is that when a node reaches some level of extreme fullness (e.g., 95%), we should stop allowing **any** writes to it so it can't possibly grow disk usage any more - and we leave a bit of disk space just to support one thing - tablet migration to newly added nodes. If I understand you correctly, what will finally bring us back below 95% will be tablets migrating out, and nothing else (not compactions, not memtable flushes, nothing).

What makes this choice viable is our new **tablet** feature. In the past it was not possible to move some of the data out of a node without allowing compactions to proceed (it was "cleanup compaction" which removed the data no longer belonging to this node). Today tablet migration can outright delete entire sstables from a node, and it doesn't require compactions to be in progress.

This choice assumes we can always solve the out-of-disk problem by adding nodes, which is true on a cloud. It is not true on an on-prem installation which wants to hover around 80%-90% disk usage, but I don't know how many users like that we actually have, and we don't need to solve their problem right now.

**Comment by denesb** [comment]
Yes, all solutions proposed here assume tablets and additional disk added by scale-out. All proposed solutions revolve around speeding up tablet migration (out of the node) and managing remaining disk space to survive the scale-out.
Yes, we do want to hover around 80%-90% disk usage and we only want to take action when disk usage is >90% (dangerously so). No need to take action at 90% yet, as we want to make it normal for a node to be at that usage level.
There are a lot of arguments around what (if anything) we need to actually do and when, as disk space runs out. We don't yet have consensus on that.

**Comment by bhalevy** [comment]
@paszkow anything planned to be done here for 2026.1.0?
Or should we move this to the backlog?

**Comment by paszkow** [comment]
> [@paszkow](https://github.com/paszkow) anything planned to be done here for 2026.1.0? Or should we move this to the backlog?

We move it to the backlog. All sub-issues were mentioned in the requirements document for the OOS, and decided to put them aside.
